### DISCUSSION TITLE
Allow replacement of S3 Bucket URL with a CDN URL

### DISCUSSION THREAD

**Discussion by Andersang** [directus/directus#20791] [Feature Requests]
### Summary

Some S3-compatible providers, such as Backblaze B2, allow for delivering of files through CDNs where the bucket URL can simply be replaced by the CDN URL. An additional environment variable to allow replacing the bucket URL with a CDN URL would be a great addition.
This is briefly mentioned in #7612 , where it was mentioned that a delivering content directly from cloud object storage is difficult to implement, but a URL replacement in this case might be simpler.

### Basic Example

Some S3-compatible providers allow for delivering of files through CDNs. One such example is Backblaze B2, which has [articles](https://www.backblaze.com/docs/cloud-storage-deliver-public-backblaze-b2-content-through-cloudflare-cdn) on how that can be done. If a B2 Bucket is successfully connected to Cloudflare, a file at `https://photos.mysite.com/file/xxx/img.jpg` (CDN URL) will be mapped to `https://f000.backblazeb2.com/file/xxx/img.jpg` (B2 Bucket URL). This remapping seems to be a simple replacement of only the front part of the full URL. 

It might be possible to allow for an additional environment variable `STORAGE_*_CDN` to route file retrieval from the CDN instead. If this environment variable is available, Directus should retrieve files from the CDN instead of the actual bucket endpoint. Uploads however, still have to be done through the bucket endpoint. Something similar has been implemented in the Wordpress plugin [WP Offload Media Lite](https://wordpress.org/plugins/amazon-s3-and-cloudfront/).

### Motivation

A CDN may deliver files at a faster speed and lower cost than the bucket used. In the case of Backblaze B2, delivery of files via the Cloudflare CDN is free, and does not count towards egress charges. This can help users of Directus save costs and enjoy faster file loading speeds.

### Detailed Design

Directus current retrieves files directly from the endpoint defined in the `STORAGE_*_ENDPOINT` environment variable. Retrieval urls seem to be constructed from this environment variable and the `filename (disk)` entry stored in the database. It might be possible to allow for an additional environment variable `STORAGE_*_CDN` to route file retrieval from the CDN instead.

The user will have to setup the link between the CDN and their storage provider on their own.

### Must Have:
- A way to set an environment variable to use a file delivery CDN

### Drawbacks

Likely no drawbacks at the moment. 

### Alternatives

I'm not sure if there is any alternative to this, considering that the file storage system in Directus seems to already be setup for this to work. 

### Adoption Strategy

Not a breaking change. Only the methods of constructing the file retrieval URLs will have to be changed.

**Comment by mattbisme** [comment]
I've absolutely been searching for an answer to this same issue. Our biggest performance bottleneck has been asset retrieval. When a visitor loads an asset, it's going from bucket to Directus instance and then finally to the user. For larger assets, this creates a noticeable delay. A direct line between the user and the CDN would provide massive gains. 

We are considering the purchase of additional storage directly attached to the instance simply for the purpose of improving performance. B2 has great pricing and performance but, with Directus as the middleman, we don't get to have everything.

**Comment by br41nslug** [comment]
> If a B2 Bucket is successfully connected to Cloudflare, a file at https://photos.mysite.com/file/xxx/img.jpg (CDN URL) will be mapped to https://f000.backblazeb2.com/file/xxx/img.jpg (B2 Bucket URL). This remapping seems to be a simple replacement of only the front part of the full URL.

Storing this URL in a custom field is something you can already achieve using a Flow or Hook extension. 

The changing the app to use arbitrary CDNs is a different story tho, i would like to see a part addressing how that would work with permission as things on CDNs are usually always publicly accessible while the files in S3 may not be.

**Reply by ggpwnkthx** [reply]
I imagine you could handle this with a flow that has a webhook trigger. You could even set up pre-signed urls using the code operations.

**Reply by orestesgaolin** [reply]
Storing a separate url for the asset is only going to work with distinct fields, but for instance it won't affect images embedded in rich text or markdown :/

**Reply by axelinternet** [reply]
Did anyone have any success with this? I'm also looking at options to access the cdn url directly without having to write too much custom code that has to be maintained

**Comment by melv1n** [comment]
I'm looking for a similiar solution. I'm using DigitalOcean Spaces and the bucket works fine with Directus. But I want to be able to give the CDN hostname in the config so the system automatically rewrites the file url (exposed via the API) to the actual CDN url. 

Would be basically be an extra entry in the docker-compose file to provide a storage_cdn variable with hostname value and it the system handles the rest itself.

**Comment by chrislsww** [comment]
Would be a great feature, indeed.

So far, I'm duplicating each image to my CDN (Bunny.net) with a flow and then write the `cdn_url` in a field for each file (which I use in the frontend). The Directus duplicate remains in my S3 object storage (I'm using Hetzner).

What would be important though: Not all files need go to the CDN, just images. 
Example: I'm also storing invoice documents, which I serve directly via Directus from the Object Storage. These are not SEO relevant and need user authentication.

So if there would be a CDN URL replacement option, I'd appreciate to add a MIME type filter and the "old way" as fallback option.

**Reply by orestesgaolin** [reply]
Would you be able to explain how are you achieving this flow? Does it also support resizing images?

**Reply by chrislsww** [reply]
Sure.

1. Create a flow in directus that triggers on "files.upload". 
2. Filter by MIME type (image). Use a condition node in this flow and use:

`{
    "$trigger": {
        "payload": {
            "type": {
                "_starts_with": "image/"
            }
        }
    }
}`

This avoid pushing other files than images to the CDN.

3. Trigger webhook in n8n (I use [n8n](https://github.com/n8n-io/n8n) for most actual actions in my automation, because they are more flexible, exportable, copyable, etc.). In n8n, I am grabbing this image via Directus API request and upload it to my CDN (I use [Bunny](https://go.lsww.de/bunny) * (affiliate link, hope that alright)). 

4. (still in the Directus this) Create an entry in a separate table called "file_names" where I have 3 fields:
* id (Index (automatically created), not really needed)
* key (the image ID)
* name (the file name)

The file name is needed to delete an image on Bunny (separate flow) and to use it in a frontend helper function to replace images in the rendered @html content, where only Directus asset URLs are provided, but not the CDN URL. The frontend helper function does the following:

1. Identify the pattern "assets/(file-id)
2. Make a request to /items/file_names 
3. Search for the key (file ID) and match with name (file name)
4. replace in rendered html content

This way, I can still use the Directus WYSIWYG editor for content writing while displaying the CDN images in the frontend.

___
Image resizing: Bunny offers a paid feature called "[Optimizer](https://bunny.net/optimizer/)" which enables you to transform images on the fly by just adding parameter to the URL. I found that simpler compared to Directus` solution because I only need to store each image once at Bunny and then just add the parameters according to my needs.
___
Hope that helps.

**Reply by orestesgaolin** [reply]
Follow up question on frontend helper. In my case I want to modify the API responses to use CDN url (both restful and graphql queries). I wonder what is the frontend helper thing you mentioned? Is that just some custom function in your frontend code or a directus-related thing?

**Reply by chrislsww** [reply]
No, that's a custom function in the frontend and has nothing to do with Directus.
